driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -316,170 +249,61 @@ private byte[] evaluate(final byte[] challenge) {
                 cachedAccessToken = null;
             }
         }
-        OidcRefreshCallback refreshCallback = getRefreshCallback();
-        if (cachedAccessToken != null) {
-            fallbackState = FallbackState.PHASE_1_CACHED_TOKEN;
-            return prepareTokenAsJwt(cachedAccessToken);
-        } else if (refreshCallback != null && cachedRefreshToken != null) {
-            assertNotNull(cachedIdpInfo);
-            // Invoke Refresh Callback using cached Refresh Token
-            validateAllowedHosts(getMongoCredential());
-            fallbackState = FallbackState.PHASE_2_REFRESH_CALLBACK_TOKEN;
-            IdpResponse result = refreshCallback.onRefresh(new OidcRefreshContextImpl(
-                    cachedIdpInfo, cachedRefreshToken, CALLBACK_TIMEOUT));
-            return populateCacheWithCallbackResultAndPrepareJwt(cachedIdpInfo, result);
-        } else {
-            // cache is empty
-
-            /*
-            A check for present idp info short-circuits phase-3a.
-
-            If a challenge is present, it can only be a response to a
-            "principal-request", so the challenge must be the resulting
-            idp info. Such a request is made during speculative auth,
-            though the source is unimportant, as long as we detect and
-            use it here.
-
-            Checking that the fallback state is not phase-3a ensures that
-            this does not loop infinitely in the case of a bug.
-             */
-            boolean idpInfoNotPresent = challenge.length == 0;
-            if (fallbackState != FallbackState.PHASE_3A_PRINCIPAL && idpInfoNotPresent) {
-                fallbackState = FallbackState.PHASE_3A_PRINCIPAL;
-                return prepareUsername(mongoCredentialWithCache.getCredential().getUserName());
-            } else {
-                IdpInfo idpInfo = toIdpInfo(challenge);
-                validateAllowedHosts(getMongoCredential());
-                IdpResponse result = requestCallback.onRequest(new OidcRequestContextImpl(idpInfo, CALLBACK_TIMEOUT));
-                fallbackState = FallbackState.PHASE_3B_REQUEST_CALLBACK_TOKEN;
-                return populateCacheWithCallbackResultAndPrepareJwt(idpInfo, result);
-            }
-        }
+        return cachedAccessToken;
     }
 
     private boolean isAutomaticAuthentication() {
         return getRequestCallback() == null;
     }
 
     private boolean clientIsComplete() {
-        return fallbackState != FallbackState.PHASE_3A_PRINCIPAL;
+        return true; // all possibilities are 1-step
```

Comment:
> It is here because it makes it easier to compare against the shouldRetryHandler method below, and the evaluate above.

Not sure I understand what this means. If this is important, could you please elaborate.

> This will have more logic in it for the human workflow, I think there is no harm in leaving it for now.

The harm is in having to track such things: https://github.com/mongodb/mongo-java-driver/pull/1134/files?show-viewed-files=true&file-filters%5B%5D=#r1493544232.